File: medium/8_String_to_Integer_(atoi)/main.cpp
#include <string>

using namespace std;

class Solution {
public:
    int myAtoi(string s) {
        int len = s.size();
        int i = 0;
        long ans = 0;
        bool foundSign = false, sign = false;
        bool firstChar = true;

        while(i < len && s[i] == ' '){
            i++;
        }

        while(i < len && (isdigit(s[i]) || (firstChar && ((s[i] == '-' && !foundSign) || (s[i] == '+' && !foundSign))))){
            if(firstChar){
                firstChar = false;
            }
            
            if(s[i] == '-'){
                foundSign = true;
                sign = true;
                i++;
                continue;
            }

            if(s[i] == '+'){
                foundSign = true;
                i++;
                continue;
            }

            if(sign){
                ans += (s[i] - '0') * -1;
            } else{
                ans += s[i] - '0';
            }

            i++;

            if(i < len && isdigit(s[i])){
                ans *= 10;
            }

            if(ans > INT_MAX){
                return INT_MAX;
            }

            if(ans < INT_MIN){
                return INT_MIN;
            }
        }

        return ans;
    }
};

File: medium/8_String_to_Integer_(atoi)/main.py
class Solution:
    def myAtoi(self, s: str) -> int:
        INT_MAX = 2 ** 31 - 1
        INT_MIN = -2 ** 31
        
        s = s.strip()
        sign = 1
        i = 0
        num = 0
        
        if not s:
            return num
        
        if s[0] == '-':
            sign = -1
            i += 1
        elif s[0] == '+':
            i += 1
        
        while i < len(s) and s[i].isdigit():
            curr_digit = ord(s[i]) - ord('0')
            
            if num > INT_MAX // 10 or (num == INT_MAX // 10 and curr_digit > 7):
                return INT_MAX if sign == 1 else INT_MIN
            
            num = num * 10 + curr_digit
            i += 1
        
        num = sign * num
        return num    